src/bundle/Controller/DownloadImageController.php

```diff
@@ -0,0 +1,258 @@
+
+    /**
+     * @param array<string, mixed> $imageMappings
+     */
+    public function __construct(
+        int $downloadLimit,
+        array $imageMappings,
+        SearchService $searchService
+    ) {
+        $this->downloadLimit = $downloadLimit;
+        $this->imageMappings = $imageMappings;
+        $this->searchService = $searchService;
+    }
+
+    /**
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\InvalidArgumentException
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\InvalidCriterionArgumentException
+     * @throws \Exception
+     */
+    public function downloadAction(string $contentIdList): Response
+    {
+        $splitContentIdList = array_map(
+            static fn (string $value): int => (int)$value,
+            explode(',', $contentIdList)
+        );
+
+        if (count($splitContentIdList) > $this->downloadLimit) {
+            throw new RuntimeException(
+                sprintf(
+                    'Total download limit in one request is %d.',
+                    $this->downloadLimit
+                )
+            );
+        }
+
+        $images = $this->loadImages($splitContentIdList);
+
+        $numberOfResults = $images->totalCount;
+        if (0 === $numberOfResults) {
+            return new Response(
+                'No results found.',
+                Response::HTTP_NOT_FOUND
+            );
+        }
+
+        if (1 === $numberOfResults) {
+            return $this->downloadSingleImage(
+                $images->getIterator()->current()->valueObject
+            );
+        }
+
+        $contentList = [];
+
+        /** @var \Ibexa\Contracts\Core\Repository\Values\Content\Search\SearchHit $image */
+        foreach ($images as $image) {
+            /** @var \Ibexa\Contracts\Core\Repository\Values\Content\Content $content */
+            $content = $image->valueObject;
+            $contentList[] = $content;
+        }
+
+        return $this->downloadArchiveWithImages($contentList);
+    }
+
+    private function downloadSingleImage(Content $content): Response
+    {
+        $value = $this->getImageValue($content);
+        $uri = $this->getImageUri($value);
+
+        $content = file_get_contents($uri);
+        if (false === $content) {
+            throw new RuntimeException(
+                sprintf(
+                    'Failed to read data from "%s"',
+                    $uri
+                )
+            );
+        }
+
+        $response = $this->createResponse(
+            $content,
+            $this->getImageFileName($value)
+        );
+
+        $response->headers->set('Content-Type', $value->mime);
+
+        return $response;
+    }
+
+    /**
+     * @param array<\Ibexa\Contracts\Core\Repository\Values\Content\Content> $contentList
+     */
+    private function downloadArchiveWithImages(array $contentList): Response
+    {
+        $archiveName = sprintf(
+            self::ARCHIVE_NAME_PATTERN,
+            time()
```

Comment:
If this is unlinked just after that, maybe we could introduce something more random than `time()`? Not sure if this is a problem tho.

Reply:
Fixed: https://github.com/ibexa/admin-ui/pull/1271/commits/066e99595489c1984d9b3bed24a204d952ad0ed0